### PR TITLE
refactor: add `SharedContext` to `Stratifier`

### PR DESCRIPTION
**Pull request by jaschdoc**
- [x] Add new separate `SharedContext[T]`
- [x] Introduce in `Stratifier`
- [x] Refactor error cases to add to `SharedContext`
- [x] Update those functions to **not** return `Validation`

Related to https://github.com/flix/flix/issues/8019

### CODE REVIEW

**Comment by magnus-madsen** [comment]
We should not have a generic `SharedContext` -- the shared part of the name refers to it being shared across threads, not across phases, but I see how that can be confusing :)

Instead, we should have `SharedContext` be declared inside `Stratifier` itself, privately and at the very bottom.

**Comment by jaschdoc** [comment]
Ahh I only did it to reduce code duplication. The name makes sense hehe :)

File: main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala [hunk 1]
```diff
@@ -86,13 +91,13 @@ object Stratifier {
...
     */
-  private def visitInstance(i0: TypedAst.Instance)(implicit root: Root, g: LabelledPrecedenceGraph, flix: Flix): Validation[TypedAst.Instance, StratificationError] =
+  private def visitInstance(i0: TypedAst.Instance)(implicit root: Root, g: LabelledPrecedenceGraph, flix: Flix, sctx: SharedContext): Validation[TypedAst.Instance, StratificationError] =
     mapN(traverse(i0.defs)(d => visitDef(d)))(ds => i0.copy(defs = ds))
 
   /**
     * Performs stratification of the given definition `def0`.
     */
-  private def visitDef(def0: Def)(implicit root: Root, g: LabelledPrecedenceGraph, flix: Flix): Validation[Def, StratificationError] =
+  private def visitDef(def0: Def)(implicit root: Root, g: LabelledPrecedenceGraph, flix: Flix, sctx: SharedContext): Validation[Def, StratificationError] =
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:100]: Sorry, one more nitty details. We always try to pass implicits in this order:

SpecificThing, LocalContext, SharedContext, root, flix

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:100]: It goes from most specific to least specific

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala:100]: Roger

**Comment by jaschdoc** [comment]
Yes :)

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**